src/expressions/method-call-expr.md

```diff
@@ -14,38 +14,68 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
 # assert!(1.14 < log_pi && log_pi < 1.15)
 ```
 
-When resolving method calls on an expression of type `A`, Rust will use the
-following order, only looking at methods that are [visible]. If the type of `A`
-is a type parameter or `Self` in a trait definitition then steps 2-4 first
-consider traits from bounds on the type paramter, then the traits that are in
-scope. For other types, only the traits that are in scope are considered.
-
-1. Inherent methods, with receiver of type `A`, `&A`, `&mut A`.
-1. Trait methods with receiver of type `A`.
-1. Trait methods with receiver of type `&A`.
-1. Trait methods with receiver of type `&mut A`.
-1. If it's possible, Rust will then repeat steps 1-5 with
-  `<A as std::ops::Deref>::Target`, and insert a dereference operator.
-1. If `A` is now an [array] type, then repeat steps 1-4 with the corresponding
-  slice type.
-
-Note: In steps 1-4, the receiver is used, not the type of `Self` nor the
-type of `A`. For example:
-
-```rust,ignore
-// `Self` is `&A`, receiver is `&A`.
-impl<'a> Trait for &'a A {
-    fn method(self) {}
+When resolving method calls on an expression of type `A`, Rust looks up methods
+both on the type itself and the traits in implements. Additionally, unlike with
+non-method function calls, the `self` parameter is special and may be
+automatically dereferenced in order to resolve it. Rust uses the following
+process to resolve method calls.
+
+First, Rust will attempt to build a list of candidate receiver types. It obtains
+these by repeatedly [dereferencing][dereference] the type, adding each type
+encountered to the list, then finally attempting an [unsized coercion] at the
```

Comment:
I'll verify this and then update if needed.

Reply:
Looks like even a CoerceUnsized coercion can work:

https://play.rust-lang.org/?gist=a08eb003e02c3f9abc0771cd7017b794&version=stable

Reply:
You're going from `&[i32; 3]` to `&[i32]`, which is exactly the case that I'm saying works.

Reply:
Err, right. Niko said it could be any unsized conversion, but I guess `CoerceUnsized` won't ever happen here, because it'll just get dereferenced and the underlying conversion applied. Likewise I think supertrait object conversion is impossible since it'll just find the method on the subtrait. The last option is unsized struct conversion, but that seems like it can't be applied here: https://play.rust-lang.org/?gist=1aa827380405e1a287b275fe0f83b69a&version=stable

Is that example correct to demonstrate that the unsized struct conversion doesn't work here? @nikomatsakis, is this the expected behaviour?